File: app/src/main/java/com/cn29/aac/di/AppModule.java
package com.cn29.aac.di;

import android.app.Application;
import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import dagger.Module;
import dagger.Provides;
import javax.inject.Singleton;

/**
 * Created by Charles Ng on 27/9/2017.
 */


@Module
public class AppModule {

  @Singleton
  @Provides
  Context provideContext(Application application) {
    return application;
  }

  @Singleton
  @Provides
  SharedPreferences provideDefaultSharedPreferences(Application application) {
    return PreferenceManager.getDefaultSharedPreferences(application);
  }

}

File: app/src/main/java/com/cn29/aac/di/AppModule.kt
package com.cn29.aac.di

import android.app.Application
import android.content.Context
import android.content.SharedPreferences
import android.preference.PreferenceManager
import com.cn29.aac.repo.util.RateLimiter
import dagger.Module
import dagger.Provides
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import java.util.concurrent.TimeUnit
import javax.inject.Singleton

/**
 * Created by Charles Ng on 27/9/2017.
 */
@Module
class AppModule {
    @Singleton
    @Provides
    fun provideContext(application: Application): Context {
        return application
    }

    @Singleton
    @Provides
    fun provideDefaultSharedPreferences(application: Application): SharedPreferences {
        return PreferenceManager.getDefaultSharedPreferences(application)
    }

    @Provides
    fun provideDispatcher(): CoroutineDispatcher {
        return Dispatchers.IO
    }

    @Provides
    fun provideRateLimiter(): RateLimiter<String> {
        return RateLimiter(
                10,
                TimeUnit.MINUTES)
    }


}